### PR TITLE
FunctionFromFunctionObjects class: wraps a vector of std::function objects into a Function object

### PR DESCRIPTION
**Pull request by luca-heltai**

### CODE REVIEW

**Comment by bangerth** [comment]
You just reinvented `Functions::ScalarFunctionFromFunctionObject` :-) 

There are a number of similar classes in namespace `Function` that would probably have done what you wanted. Can you take a look? I'm pretty sure that most cases there are all you need ;-)

**Comment by tamiko** [comment]
A quick question:

Have you seen the three function classes `ScalarFunctionFromFunctionObject` and `VectorFunctionFromFunctionObject` and `VectorFunctionFromTensorFunction`? Is there a possiblity to clean up our codebase a bit and retire the other three (best case would be to have drop-in replacements with `LambdaFunction`)?

**Comment by bangerth** [comment]
On 4/12/19 2:34 PM, Matthias Maier wrote:
> Have you seen the three function classes 
> |ScalarFunctionFromFunctionObject| and 
> |VectorFunctionFromFunctionObject| and 
> |VectorFunctionFromTensorFunction|? Is there a possiblity to clean up 
> our codebase a bit and retire the other three (best case would be to 
> have drop-in replacements with |LambdaFunction|)?

I don't think that the word "Lambda" in the name is actually helpful, 
because you can initialize these objects with a lot of other things that 
are not lambda functions.

**Comment by luca-heltai** [comment]
Ah! That's interesting... they do not show up here

https://www.dealii.org/developer/doxygen/deal.II/group__functions.html

 and I was not aware of their existance... :) 

The only thing I don't see, is how to construct quickly a genuine vector function object (which the method I provide here give by passing a vector of std::functions), but yes, this is just reinventing the weel... 

If you don't like the name, we could simply make one single FunctionFromFunctionObject (modifying the existing one by simply adding a constructor to it) that allows one to construct a scalar or vector function according to how many std::function objects you pass.

**Comment by bangerth** [comment]
These classes are all pretty small. I'd just add a `VectorFunctionFromFunctionObjects` that does essentially what you are already doing, rather than trying to merge all of these cases into one class that tries to do it all.

File: include/deal.II/base/function.h [hunk 1]
```diff
@@ -825,6 +826,135 @@ class VectorFunctionFromScalarFunctionObject
...
+  FunctionFromFunctionObjects(
+    const std::vector<std::function<RangeNumberType(const Point<dim> &)>>
+      &          values,
+    const double initial_time = 0.0);
+
+  /**
+   * Constructor for functions where we know both values and gradients.
+   *
+   * The resulting function will have a number of components equal to the size
+   * of the vector @p values. If the size of @p values and @p gradients do not
```
Comment (masterleinad) [inline include/deal.II/base/function.h]:
```suggestion
   * of the vector @p values. If the size of @p values and @p gradients does not
```

File: include/deal.II/base/function.h [hunk 2]
```diff
@@ -827,6 +828,143 @@ class VectorFunctionFromScalarFunctionObject
...
+ * To be more concrete, let us consider the following example:
+ * @code
+ * RangeNumberType
+ * first_component(const Point<2> &p)
+ * {
+ *   return 1.0;
+ * }
+ *
+ * RangeNumberType
+ * first_component(const Point<2> &p)
```
Comment (masterleinad) [inline include/deal.II/base/function.h]:
```suggestion
 * second_component(const Point<2> &p)
```

File: include/deal.II/base/function.h [hunk 3]
```diff
@@ -827,6 +828,143 @@ class VectorFunctionFromScalarFunctionObject
...
+public:
+  /**
+   * Default constructor.
+   *
+   * This constructor does not initialize the internal methods. To have a
+   * usable function, you need to call at least the set_function_values()
+   * method. If you need also the gradients of the solution, then you must
+   * also call the set_function_gradients() method.
+   */
+  FunctionFromFunctionObjects(const unsigned int n_components = 1,
```
Comment (tjhei) [inline include/deal.II/base/function.h]: This should be explicit

File: include/deal.II/base/function.templates.h [hunk 4]
```diff
@@ -783,6 +783,93 @@ VectorFunctionFromTensorFunction<dim, RangeNumberType>::vector_value_list(
...
+
+template <int dim, typename RangeNumberType>
+RangeNumberType
+FunctionFromFunctionObjects<dim, RangeNumberType>::value(
+  const Point<dim> & p,
+  const unsigned int component) const
+{
+  AssertIndexRange(component, this->n_components);
+  Assert(function_values[component],
+         ExcMessage("Empty lambda for component values"));
```
Comment (tjhei) [inline include/deal.II/base/function.templates.h]: would an error message like "Accessing value() in FunctionFromFunctionObjects requires setting the std::function objects for the value" or something like that be better?

File: include/deal.II/base/function.templates.h [hunk 5]
```diff
@@ -783,6 +783,93 @@ VectorFunctionFromTensorFunction<dim, RangeNumberType>::vector_value_list(
...
+
+template <int dim, typename RangeNumberType>
+Tensor<1, dim, RangeNumberType>
+FunctionFromFunctionObjects<dim, RangeNumberType>::gradient(
+  const Point<dim> & p,
+  const unsigned int component) const
+{
+  AssertIndexRange(component, this->n_components);
+  Assert(function_gradients[component],
+         ExcMessage("Empty lambda for component gradients"));
```
Comment (tjhei) [inline include/deal.II/base/function.templates.h]: similarly...

**Pull request merged**